codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/DafnyToSmithyShapeVisitor.java

```diff
@@ -349,17 +350,59 @@ public String doubleShape(DoubleShape shape) {
-                dataSource);
+
+        String returnString = """
+                func() %s {
+                """.formatted(context.symbolProvider().toSymbol(shape));
+        returnString += """
+                    if %s == nil {
+                        return nil
+                    }
+                    var union %s
+                """.formatted(
+                    dataSource,
+                    context.symbolProvider().toSymbol(shape)
+                );
+        for (var member : shape.getAllMembers().values()) {
+            Shape targetShape = context.model().expectShape(member.getTarget());
+            String memberName = context.symbolProvider().toMemberName(member);
+            String rawUnionDataSource = "(" + dataSource + ".(" + DafnyNameResolver.getDafnyType(shape, context.symbolProvider().toSymbol(shape)) + "))";
+            // unwrap union type, assert it then convert it to its member type with Dtor_ (example: Dtor_BlobValue()). unionDataSource is not a wrapper object until now.
+            String unionDataSource = rawUnionDataSource + ".Dtor_" + memberName.replace(shape.getId().getName() + "Member", "") + "()";
+            Boolean isMemberShapePointable = GoPointableIndex.of(context.model()).isPointable(targetShape) && GoPointableIndex.of(context.model()).isDereferencable(targetShape);
+            String pointerForPointableShape = isMemberShapePointable ? "*" : "";
+            returnString += """
+                        if ((%s).%s()) {
+                    """.formatted(
+                        rawUnionDataSource,
+                        memberName.replace(shape.getId().getName() + "Member", "Is_")
+                        );
+            if (!(targetShape.isListShape() || targetShape.isStructureShape())) {
+                // All other shape except list and structure needs a Wrapper object but unionDataSource is not a Wrapper object. 
+                returnString += """
+                    var dataSource = Wrappers.Companion_Option_.Create_Some_(%s)
+                    """.formatted(
+                        unionDataSource
+                    );
+                unionDataSource = "dataSource.UnwrapOr(nil)";
+            }
+            returnString += """
+                            union = &%s.%s{
+                                    Value: %s(%s),
+                                }
+                            }
+                        """.formatted(
+                            SmithyNameResolver.smithyTypesNamespace(shape),
+                            memberName,
+                            pointerForPointableShape,
+                            targetShape.accept(
+                                    new DafnyToSmithyShapeVisitor(context, unionDataSource, writer, isConfigShape, isMemberShapePointable)
+                                ));
+        }
+        returnString += """
+            return union
+            }()""";
+        return returnString;
```

Comment:
I'd recommend to write things in `"""` blocks and then use `%s`. This makes it more readable over using `+` all over the place.

Reply:
Done